File: app/src/main/java/com/wolcano/musicplayer/music/utils/PermissionUtils.java
package com.wolcano.musicplayer.music.utils;

import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Build;
import android.util.SparseArray;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;


public class PermissionUtils {

    public interface PermInterface {
        void onPermGranted();

        void onPermUnapproved();
    }

    private static Set<String> permSets;
    private PermInterface permInterface;
    private static AtomicInteger atomicInteger = new AtomicInteger(0);
    private Object object;
    private static SparseArray<PermInterface> sparseArray = new SparseArray<>();

    private PermissionUtils(Object object) {
        this.object = object;
    }

    public static PermissionUtils with(@NonNull Activity activity) {
        return new PermissionUtils(activity);
    }

    public static PermissionUtils with(@NonNull Fragment fragment) {
        return new PermissionUtils(fragment);
    }

    private String[] perms;

    public PermissionUtils permissions(@NonNull String... perms) {
        this.perms = perms;
        return this;
    }

    public PermissionUtils result(@Nullable PermInterface sonuc) {
        this.permInterface = sonuc;
        return this;
    }

    public void reqPerm() {
        Activity act = getAct(object);
        if (act == null) {
            throw new IllegalArgumentException("permission error");
        }
        setPerms(act);
        for (String permission : perms) {
            if (!permSets.contains(permission)) {
                if (permInterface != null) {
                    permInterface.onPermUnapproved();
                }
                return;
            }
        }
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            if (permInterface != null) {
                permInterface.onPermGranted();
            }
            return;
        }

        List<String> unapprovedPermsList = getUnapprovedPerms(act, perms);
        if (unapprovedPermsList.isEmpty()) {
            if (permInterface != null) {
                permInterface.onPermGranted();
            }
            return;
        }
        int reqCode = genRequestCode();
        String[] unapprovedPerms = unapprovedPermsList.toArray(new String[unapprovedPermsList.size()]);
        reqPerms(object, unapprovedPerms, reqCode);
        sparseArray.put(reqCode, permInterface);
    }

    public static void onRequestPermissionsResult(int requestCode, @NonNull String[] perms, @NonNull int[] grantResults) {
        PermInterface sonuc = sparseArray.get(requestCode);

        if (sonuc == null) {
            return;
        }

        sparseArray.remove(requestCode);

        for (int grantResult : grantResults) {
            if (grantResult != PackageManager.PERMISSION_GRANTED) {
                sonuc.onPermUnapproved();
                return;
            }
        }
        sonuc.onPermGranted();
    }

    @TargetApi(Build.VERSION_CODES.M)
    private static void reqPerms(Object object, String[] permissions, int requestCode) {
        if (object instanceof Activity) {
            ((Activity) object).requestPermissions(permissions, requestCode);
        } else if (object instanceof Fragment) {
            ((Fragment) object).requestPermissions(permissions, requestCode);
        }
    }

    private static List<String> getUnapprovedPerms(Context context, String[] perms) {
        List<String> unapprovedPermsList = new ArrayList<>();
        for (String perm : perms) {
            if (ContextCompat.checkSelfPermission(context, perm) != PackageManager.PERMISSION_GRANTED) {
                unapprovedPermsList.add(perm);
            }
        }
        return unapprovedPermsList;
    }

    private static synchronized void setPerms(Context context) {
        if (permSets == null) {
            permSets = new HashSet<>();
            try {
                PackageInfo pckgInfo = context.getPackageManager().getPackageInfo(context.getPackageName(), PackageManager.GET_PERMISSIONS);
                String[] perms = pckgInfo.requestedPermissions;
                Collections.addAll(permSets, perms);
            } catch (PackageManager.NameNotFoundException e) {
                e.printStackTrace();
            }
        }
    }

    private static Activity getAct(Object object) {
        if (object != null) {
            if (object instanceof Activity) {
                return (Activity) object;
            } else if (object instanceof Fragment) {
                return ((Fragment) object).getActivity();
            }
        }
        return null;
    }

    private static int genRequestCode() {
        return atomicInteger.incrementAndGet();
    }
}

File: app/src/main/java/com/wolcano/musicplayer/music/utils/PermissionUtils.kt
package com.wolcano.musicplayer.music.utils

import android.annotation.TargetApi
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.util.SparseArray
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import java.util.*
import java.util.concurrent.atomic.AtomicInteger

class PermissionUtils private constructor(private val `object`: Any) {

    interface PermInterface {
        fun onPermGranted()
        fun onPermUnapproved()
    }

    private var permInterface: PermInterface? = null

    private lateinit var perms: Array<out String>

    fun permissions(vararg perms: String): PermissionUtils {
        this.perms = perms
        return this
    }

    fun result(result: PermInterface?): PermissionUtils {
        permInterface = result
        return this
    }

    fun requestPermissions() {
        val act = getAct(
            `object`
        ) ?: throw IllegalArgumentException("permission error")
        setPerms(act)
        for (permission in perms) {
            if (!permSets!!.contains(permission)) {
                if (permInterface != null) {
                    permInterface!!.onPermUnapproved()
                }
                return
            }
        }
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            permInterface?.onPermGranted()
            return
        }
        val unapprovedPermsList = getUnapprovedPerms(act, perms)
        if (unapprovedPermsList.isEmpty()) {
            permInterface?.onPermGranted()
            return
        }
        val reqCode = genRequestCode()
        val unapprovedPerms = unapprovedPermsList.toTypedArray()
        requestPermissionss(`object`, unapprovedPerms, reqCode)
        sparseArray.put(reqCode, permInterface)
    }

    companion object {
        private var permSets: MutableSet<String>? = null
        private val atomicInteger = AtomicInteger(0)
        private val sparseArray = SparseArray<PermInterface?>()
        fun with(activity: Activity): PermissionUtils {
            return PermissionUtils(activity)
        }

        fun with(fragment: Fragment): PermissionUtils {
            return PermissionUtils(fragment)
        }

        fun onRequestPermissionsResult(
            requestCode: Int,
            perms: Array<String?>,
            grantResults: IntArray
        ) {
            val result = sparseArray[requestCode]
                ?: return
            sparseArray.remove(requestCode)
            for (grantResult in grantResults) {
                if (grantResult != PackageManager.PERMISSION_GRANTED) {
                    result.onPermUnapproved()
                    return
                }
            }
            result.onPermGranted()
        }

        @TargetApi(Build.VERSION_CODES.M)
        private fun requestPermissionss(`object`: Any, permissions: Array<String>, requestCode: Int) {
            if (`object` is Activity) {
                `object`.requestPermissions(permissions, requestCode)
            } else if (`object` is Fragment) {
                `object`.requestPermissions(permissions, requestCode)
            }
        }

        private fun getUnapprovedPerms(context: Context, perms: Array<out String>): List<String> {
            val unapprovedPermsList: MutableList<String> = ArrayList()
            for (perm in perms) {
                if (ContextCompat.checkSelfPermission(
                        context,
                        perm
                    ) != PackageManager.PERMISSION_GRANTED
                ) {
                    unapprovedPermsList.add(perm)
                }
            }
            return unapprovedPermsList
        }

        @Synchronized
        private fun setPerms(context: Context) {
            if (permSets == null) {
                permSets = HashSet()
                try {
                    val packageInfo = context.packageManager.getPackageInfo(
                        context.packageName,
                        PackageManager.GET_PERMISSIONS
                    )
                    val permissions = packageInfo.requestedPermissions
                    Collections.addAll(permSets, *permissions)
                } catch (e: PackageManager.NameNotFoundException) {
                    e.printStackTrace()
                }
            }
        }

        private fun getAct(`object`: Any?): Activity? {
            if (`object` != null) {
                if (`object` is Activity) {
                    return `object`
                } else if (`object` is Fragment) {
                    return `object`.activity
                }
            }
            return null
        }

        private fun genRequestCode(): Int {
            return atomicInteger.incrementAndGet()
        }
    }

}